File: src/test_lee_rclpy/test_lee_rclpy/coordinates/position.py

import random

from msg_srv_interface_py_test.msg import Coordinates
from rcl_interfaces.msg import SetParametersResult
import rclpy
from rclpy.node import Node
from rclpy.parameter import  Parameter
from rclpy.qos import QoSDurabilityPolicy
from rclpy.qos import QoSHistoryPolicy
from rclpy.qos import QoSProfile
from rclpy.qos import QoSReliabilityPolicy


class Position(Node):

    def __init__(self):
        super().__init__('position')
        self.declare_parameter('qos_depth', 10)
        qos_depth = self.get_parameter('qos_depth').value
        self.declare_parameter('min_random_num', 10)
        self.min_random_num = self.get_parameter('min_random_num').value
        self.declare_parameter('max_random_num', 20)
        self.max_random_num = self.get_parameter('max_random_num').value
        self.add_on_set_parameters_callback(self.update_parameter)

        QOS_RKL10V = QoSProfile(
            reliability=QoSReliabilityPolicy.RELIABLE,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=qos_depth,
            durability=QoSDurabilityPolicy.VOLATILE)

        self.coordinate_publisher = self.create_publisher(
            Coordinates,
            'position',
            QOS_RKL10V)

        self.timer = self.create_timer(1.0, self.publish_random_coordinates)

    def publish_random_coordinates(self):
        msg = Coordinates()
        msg.stamp = self.get_clock().now().to_msg()
        msg.coordinate_x = int(random.randint(self.min_random_num, self.max_random_num))
        msg.coordinate_y = int(random.randint(self.min_random_num, self.max_random_num))
        self.coordinate_publisher.publish(msg)
        self.get_logger().info('Published time: {0}'.format(msg.stamp))
        self.get_logger().info('Published coordinates_x: {0}'.format(msg.coordinate_x))
        self.get_logger().info('Published coordinates_y: {0}'.format(msg.coordinate_y))

    def update_parameter(self, params):
        for param in params:
            if param.name == 'min_random_num' and param.type_ == Parameter.Type.INTEGER:
                self.min_random_num = param.value
            elif param.name == 'max_random_num' and param.type_ == Parameter.Type.INTEGER:
                self.max_random_num = param.value
        return SetParametersResult(successful=True)


def main(args=None):
    rclpy.init(args=args)
    try:
        position = Position()
        try:
            rclpy.spin(position)
        except KeyboardInterrupt:
            position.get_logger().info('Keyboard Interrupt (SIGINT)')
        finally:
            position.destroy_node()
    finally:
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/test_lee_rclcpp/src/coordinates/position.cpp
#include <cstdio>
#include <memory>
#include <string>
#include <utility>
#include <random>

#include "rclcpp/rclcpp.hpp"
#include "rcutils/cmdline_parser.h"

#include "coordinates/position.hpp"

using namespace std::chrono_literals;

Position::Position(const rclcpp::NodeOptions & node_options)
: Node("position", node_options),
  min_random_num_(0.0),
  max_random_num_(0.0)
{
  this->declare_parameter("qos_depth", 10);
  int8_t qos_depth = this->get_parameter("qos_depth").get_value<int8_t>();
  this->declare_parameter("min_random_num", 0.0);
  min_random_num_ = this->get_parameter("min_random_num").get_value<float>();
  this->declare_parameter("max_random_num", 9.0);
  max_random_num_ = this->get_parameter("max_random_num").get_value<float>();
  this->update_parameter();

  const auto QOS_RKL10V =
    rclcpp::QoS(rclcpp::KeepLast(qos_depth)).reliable().durability_volatile();

  coordinates_publisher_ =
    this->create_publisher<Coordinates>("position", QOS_RKL10V);

  timer_ =
    this->create_wall_timer(1s, std::bind(&Position::publish_random_coordinates, this));
}

Position::~Position()
{
}

void Position::publish_random_coordinates()
{
  std::random_device rd;
  std::mt19937 gen(rd());
  std::uniform_real_distribution<float> distribution(min_random_num_, max_random_num_ );

  msg_srv_action_interface_cpp::msg::Coordinates msg;
  msg.stamp = this->now();
  msg.coordinate_x = distribution(gen);
  msg.coordinate_y = distribution(gen);
  coordinates_publisher_->publish(msg);
  RCLCPP_INFO(this->get_logger(), "Coordinates!!");
  RCLCPP_INFO(this->get_logger(), "Published coordinate_x: %d", msg.coordinate_x);
  RCLCPP_INFO(this->get_logger(), "Published coordinate_y: %d", msg.coordinate_y);
}

void Position::update_parameter()
{
  parameters_client_ = std::make_shared<rclcpp::AsyncParametersClient>(this);
  while (!parameters_client_->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
      return;
    }
    RCLCPP_INFO(this->get_logger(), "service not available, waiting again...");
  }

  auto param_event_callback =
    [this](const rcl_interfaces::msg::ParameterEvent::SharedPtr event) -> void
    {
      for (auto & changed_parameter : event->changed_parameters) {
        if (changed_parameter.name == "min_random_num") {
          auto value = rclcpp::Parameter::from_parameter_msg(changed_parameter).as_double();
          min_random_num_ = value;
        } else if (changed_parameter.name == "max_random_num") {
          auto value = rclcpp::Parameter::from_parameter_msg(changed_parameter).as_double();
          max_random_num_ = value;
        }
      }
    };

  parameter_event_sub_ = parameters_client_->on_parameter_event(param_event_callback);
}

void print_help()
{
  printf("For argument node:\n");
  printf("node_name [-h]\n");
  printf("Options:\n");
  printf("\t-h Help           : Print this help function.\n");
}

int main(int argc, char * argv[])
{
  if (rcutils_cli_option_exist(argv, argv + argc, "-h")) {
    print_help();
    return 0;
  }

  rclcpp::init(argc, argv);

  auto position = std::make_shared<Position>();

  rclcpp::spin(position);

  rclcpp::shutdown();

  return 0;
}
